include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_radix_sort.h

```diff
@@ -514,21 +579,22 @@ __radix_sort_reorder_submit(_ExecutionPolicy&& __exec, ::std::size_t __segments,
 
                 _PeerHelper __peer_prefix_hlp(__self_item, __peer_temp);
 
-                // 1. create a private array for storing offset values
-                //    and add total offset and offset for compute unit for a certain radix state
-                _OffsetT __offset_arr[__radix_states];
                 const ::std::size_t __scan_size = __segments + 1;
-                _OffsetT __scanned_bin = 0;
-                __offset_arr[0] = __offset_rng[__wgroup_idx];
-                for (::std::uint32_t __radix_state_idx = 1; __radix_state_idx < __radix_states; ++__radix_state_idx)
+                for (::std::uint32_t __radix_state = __self_lidx; __radix_state < __radix_states; __radix_state += __sg_size)
```

Comment:
BTW, we allow that `__sg_size`  is sub group size here, right?
But if we look on the sycl::nd_range for re-order we may notice that we have the iteration  space `(__segments * __sg_size, __sg_size)`, where `__sg_size` is a work group size and `__segments` is number of work groups.
`__sg_size` is very small  (8 or 4 for int)... So, it looks like we have low GPU occupancy here... Or I am wrong?

Reply:
This loop is filling the offset array (which is size radix states) cooperatively amongst the work-items. For the case of 4 bits, we need to read 16 elements from the input offset range. If the subgroup size is less than 16 (such as 8 or 4 in your example), then each work-item is actually performing multiple trips through this loop and the occupancy is high (in the sense that each work-item is busy). It is only in the case where the subgroup size is larger than 16 that some work-items are left idle for this loop.